Allow election of nodes outside voting config

DaveCTurner opened the pull request:
Today we suppress election attempts on master-eligible nodes that are not in
the voting configuration. In fact this restriction is not necessary: any
master-eligible node can safely become master as long as it has a fresh enough
cluster state and can gather a quorum of votes. Moreover, this restriction is
sometimes undesirable: there may be a reason why we do not want any of the
nodes in the voting configuration to become master.

The reason for this restriction is as follows. If you want to shut the master
down then you might first exclude it from the voting configuration. When this
exclusion succeeds you might reasonably expect that a new master has been
elected, since the voting config exclusion is almost always a step towards
shutting the node down. If we allow nodes outside the voting configuration to
be the master then the excluded node will continue to be master, which is
confusing.

This commit adjusts the logic to allow master-eligible nodes to attempt an
election even if they are not in the voting configuration. If such a master is
successfully elected then it adds itself to the voting configuration. This
commit also adjusts the logic that causes master nodes to abdicate when they
are excluded from the voting configuration, to avoid the confusion described
above.

Relates #37712, #37802.

DaveCTurner committed:
Allow election of nodes outside voting config

Today we suppress election attempts on master-eligible nodes that are not in
the voting configuration. In fact this restriction is not necessary: any
master-eligible node can safely become master as long as it has a fresh enough
cluster state and can gather a quorum of votes. Moreover, this restriction is
sometimes undesirable: there may be a reason why we do not want any of the
nodes in the voting configuration to become master.

The reason for this restriction is as follows. If you want to shut the master
down then you might first exclude it from the voting configuration. When this
exclusion succeeds you might reasonably expect that a new master has been
elected, since the voting config exclusion is almost always a step towards
shutting the node down. If we allow nodes outside the voting configuration to
be the master then the excluded node will continue to be master, which is
confusing.

This commit adjusts the logic to allow master-eligible nodes to attempt an
election even if they are not in the voting configuration. If such a master is
successfully elected then it adds itself to the voting configuration. This
commit also adjusts the logic that causes master nodes to abdicate when they
are excluded from the voting configuration, to avoid the confusion described
above.

Relates #37712, #37802.

elasticmachine commented:
Pinging @elastic/es-distributed

DaveCTurner commented on server/src/main/java/org/elasticsearch/cluster/coordination/Coordinator.java:
```diff
@@ -189,7 +190,8 @@ public Coordinator(String nodeName, Settings settings, ClusterSettings clusterSe
         return new ClusterFormationState(settings, getStateForMasterService(), peerFinder.getLastResolvedAddresses(),
-            StreamSupport.stream(peerFinder.getFoundPeers().spliterator(), false).collect(Collectors.toList()), getCurrentTerm());
+            Stream.concat(Stream.of(getLocalNode()), StreamSupport.stream(peerFinder.getFoundPeers().spliterator(), false))
+                .collect(Collectors.toList()), getCurrentTerm());
```
Fixes unrelated bug in which the `ClusterFormationFailureHelper` wasn't claiming to have discovered the local node.

ywelsch replied:
can you add a test for this?

ywelsch commented on server/src/main/java/org/elasticsearch/cluster/coordination/Coordinator.java:
```diff
@@ -1297,16 +1299,16 @@ public void onSuccess(String source) {
                                                 .collect(Collectors.toList());
                                             if (masterCandidates.isEmpty() == false) {
                                                 abdicateTo(masterCandidates.get(random.nextInt(masterCandidates.size())));
                                             }
```
I would feel better about calling `scheduleReconfigurationIfNeeded()` in the else branch here, even if it turns out to be a NOOP. Otherwise the interaction between what the reconfigurator does and the logic here is too subtle.

ywelsch reviewed (approved):
Thanks for picking this up. I've left 2 nits.

DaveCTurner committed:
Merge branch 'master' into 2019-06-14-elect-nodes-outside-voting-configuration

DaveCTurner committed:
Schedule a reconfiguration if abdication wasn't possible

DaveCTurner committed:
Backout

DaveCTurner commented:
I backed out the change to the `ClusterFormationFailureHelper` in c8d8caf, moved it to #43316 and added a test.

I addressed the other comment in 9636d55.
